dnssec-tools/validator/doc/libval.3
.\" Automatically generated by Pod::Man 4.09 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.if !\nF .nr F 0
.if \nF>0 \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    if !\nF==2 \{\
.        nr % 0
.        nr F 2
.    \}
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "libval 3"
.TH libval 3 "2018-08-08" "perl v5.26.2" "Programmer's Manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
val_create_context()
val_free_context() \- manage validator context
.PP
val_context_setqflags() \- manage validator context flags
.PP
val_resolve_and_check(), val_free_result_chain() \- query and validate
answers from a DNS name server
.PP
val_istrusted() \- check if status value corresponds to that of a
trustworthy answer
.PP
val_isvalidated() \- check if status value represents an
answer that cryptographically chains down from a configured
trust anchor
.PP
val_does_not_exist() \- check if status value represents
one of the non\-existence types
.PP
p_val_status(), p_ac_status(), p_val_error() \- display validation status,
authentication chain status and error information
.PP
val_log_add_optarg \- control log message verbosity and output location
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&  #include <validator.h>
\&
\&  int val_create_context(const char *label, val_context_t **newcontext);
\&
\&  int val_create_context_ex(char *label, 
\&                            val_context_opt_t *opt, 
\&                            val_context_t ** newcontext);
\&
\&  int val_context_setqflags(val_context_t *context, 
\&                            unsigned char action, 
\&                            unsigned int flags);
\&
\&  int val_context_store_ns_for_zone(val_context_t *context, 
\&                                    char * zone, 
\&                                    char *resp_server,
\&                                    int recursive)
\&
\&  int val_resolve_and_check(val_context_t *context,
\&                         const char *domain_name,
\&                         int class,
\&                         int type,
\&                         unsigned int  flags,
\&                         struct val_result_chain  **results);
\&
\&  char *p_val_status(val_status_t valerrno);
\&
\&  char *p_ac_status(val_astatus_t auth_chain_status);
\&
\&  char *p_val_error(int err);
\&
\&  int val_istrusted(val_status_t val_status);
\&
\&  int val_isvalidated(val_status_t val_status);
\&
\&  int val_does_not_exist(val_status_t status);
\&
\&  val_log_t *val_log_add_optarg(const char *args, int use_stderr);
\&
\&  void val_free_result_chain(struct val_result_chain *results);
\&
\&  void val_free_context(val_context_t *context);
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fI\fIval_resolve_and_check()\fI\fR function queries a set of name servers for
the \fI<domain_name, type, class\fR> tuple and verifies and validates the
response.  Verification involves checking the RRSIGs, and validation is
verification of an authentication chain from a configured trust anchor.
.PP
The \fIflags\fR parameter can be used to control the results of validation.
The following values, which may be ORed together, are currently defined 
for this field:
.IP "\fB\s-1VAL_QUERY_DONT_VALIDATE\s0\fR" 4
.IX Item "VAL_QUERY_DONT_VALIDATE"
This flag causes the validator to disable validation for this query.
.IP "\fB\s-1VAL_QUERY_IGNORE_SKEW\s0\fR" 4
.IX Item "VAL_QUERY_IGNORE_SKEW"
This flag causes the validator to disable checking signature inception and
expiration times on RRSIGs.
.IP "\fB\s-1VAL_QUERY_AC_DETAIL\s0\fR" 4
.IX Item "VAL_QUERY_AC_DETAIL"
This flag causes the validator to copy the authentication chain details into
the \fIval_rc_answer\fR member within the returned \fIval_result_chain\fR structure.
.IP "\fB\s-1VAL_QUERY_ASYNC\s0\fR" 4
.IX Item "VAL_QUERY_ASYNC"
This flag enables asynchronous query resolution for that lookup.
.IP "\fB\s-1VAL_QUERY_NO_DLV\s0\fR" 4
.IX Item "VAL_QUERY_NO_DLV"
This flag causes the validator to disable \s-1DLV\s0 processing for this query.  
This is only available if the \fB\f(BIlibval\fB\|(3)\fR library has
been compiled with \s-1DLV\s0 support.
.IP "\fB\s-1VAL_QUERY_NO_EDNS0_FALLBACK\s0\fR" 4
.IX Item "VAL_QUERY_NO_EDNS0_FALLBACK"
In querying various name servers, \fB\f(BIlibsres\fB\|(3)\fR will also attempt multiple
\&\s-1EDNS0\s0 sizes, ending with a query that has \s-1EDNS0\s0 disabled (i.e. no \s-1CD\s0 bit set).
This flag causes \fBlibval\fR to disable \s-1EDNS0\s0 fallback for the query.
.IP "\fB\s-1VAL_QUERY_ITERATE\s0\fR" 4
.IX Item "VAL_QUERY_ITERATE"
This flag forces \fBlibval\fR to iteratively lookup the query by querying
various name servers in the delegation hierarchy, instead of
requesting this information from
any caching name server that may be configured in \fBdnsval.conf\fR.
.IP "\fB\s-1VAL_QUERY_SKIP_RESOLVER\s0\fR" 4
.IX Item "VAL_QUERY_SKIP_RESOLVER"
This flag forces \fBlibval\fR to only look at its cache while trying to resolve a
name.
.IP "\fB\s-1VAL_QUERY_SKIP_CACHE\s0\fR" 4
.IX Item "VAL_QUERY_SKIP_CACHE"
There are two levels of caching that happen within libval: at the
assertion level and at the answer level. This flag forces \fBlibval\fR to
ignore data in both caches while trying to resolve a name.
.IP "\fB\s-1VAL_QUERY_SKIP_ANS_CACHE\s0\fR" 4
.IX Item "VAL_QUERY_SKIP_ANS_CACHE"
There are two levels of caching that happen within libval: at the
assertion level and at the answer level. This flag forces \fBlibval\fR to
ignore data in the answer cache while trying to resolve a name.
If an assertion already exists for the given name,class and type that
assertion is re-used till it expires. Once it expires the name is looked
up via a query even if an newer record (fetched by another assertion in
the same or different context) is available in its answer cache.
.PP
The first parameter to \fI\fIval_resolve_and_check()\fI\fR is the validator context.
Applications can create a new validator context using the
\&\fI\fIval_create_context()\fI\fR function.  This function parses the resolver and
validator configuration files and creates the \fInewcontext\fR handle for this
parsed information.  Information stored as part of the validator context
includes the validation policy the and resolver policy.
.PP
Validator and resolver policies are read from the \fB/etc/dnsval.conf\fR and
\&\fB/etc/resolv.conf\fR files by default.  \fB/etc/root.hints\fR provides
bootstrapping information for the validator when it functions as a full
resolver (see \fB\f(BIdnsval.conf\fB\|(3)\fR). These defaults may be different if 
any other value was specified at library configure time.  If the default 
resolver configuration file is not found at the specified location, \fBlibval\fR
will also try to fall back to \fB/etc/resolv.conf\fR as a last resort.
.PP
Applications may also create a validator context with a custom policy 
using the \fI\fIval_create_context_ex()\fI\fR function.
.PP
The \fIopt\fR argument in \fI\fIval_create_context_ex()\fI\fR points to the
following structure:
.PP
.Vb 10
\&    typedef struct val_context_opt {
\&        unsigned int vc_qflags;
\&        unsigned int vc_polflags;
\&        char *vc_valpol;
\&        char *vc_nslist;
\&        char *vc_val_conf;
\&        char *vc_res_conf;
\&        char *vc_root_conf;
\&        val_global_opt_t *gopt, 
\&    } val_context_opt_t;
.Ve
.PP
\&\fIvc_qflags\fR specify the default query flags for queries associated with
the context. This field is identical to the \fIflags\fR argument in the
\&\fI\fIval_resolve_and_check()\fI\fR function.
.PP
The \fIvc_val_conf\fR, \fIvc_res_conf\fR, and \fIvc_root_conf\fR fields can be used to
specify alternative configuration files in place of the default files used for
the validator, resolver, and root hints configuration information.
.PP
The \fIvc_valpol\fR field can be used to supply a custom validator
configuration \*(L"snippet\*(R" to the validator. The validator parses this
information and applies it to the default validator configuration either
as an extension or as an override, as specified by the \fIvc_polflags\fR
field. The \fIvc_nslist\fR field, similarly, can be used to supply a custom
list of recursive name servers to send queries to, in addition (or in
place of) the list of name servers specified in the \fBresolv.conf\fR file.
The list of name servers in \fIvc_res_conf\fR must be specified as a
white-space delimited list of \s-1IP\s0 or IPv6 addresses.
.PP
A custom mapping between a zone and a name server may be created
using the \fB\f(BIval_context_store_ns_for_zone()\fB\fR function.
.PP
\&\fIvc_polflags\fR can be configured with the following values:
.IP "\fB\s-1CTX_DYN_POL_VAL_OVR\s0\fR" 4
.IX Item "CTX_DYN_POL_VAL_OVR"
When this flag is set only the validator configuration specified in the
\&\fIvc_valpol\fR field is used for the validator configuration.  Options
specified in the \fBdnsval.conf\fR file are ignored.
.IP "\fB\s-1CTX_DYN_POL_RES_OVR\s0\fR" 4
.IX Item "CTX_DYN_POL_RES_OVR"
When this flag is set only the name servers specified in the
\&\fIvc_nslist\fR field is used for the resolver configuration.  Options
specified in the \fBresolv.conf\fR file are ignored.
.IP "\fB\s-1CTX_DYN_POL_GLO_OVR\s0\fR" 4
.IX Item "CTX_DYN_POL_GLO_OVR"
When this flag is set only the global options set in \fIgopt\fR field are
used.  Global options specified in the \fBdnsval.conf\fR file are ignored.
.IP "\fB\s-1CTX_DYN_POL_RES_NRD\s0\fR" 4
.IX Item "CTX_DYN_POL_RES_NRD"
When this flag is set outgoing queries have the Recursion Desired (\s-1RD\s0)
bit set to 0.  This is useful if queries need to be sent to an
authoritative-only name server.
.PP
The \fIgopt\fR field points to the following structure:
.PP
.Vb 11
\&    typedef struct val_global_opt {
\&        int local_is_trusted;
\&        long edns0_size;
\&        int env_policy;
\&        int app_policy;
\&        char *log_target;
\&        int closest_ta_only;
\&        int rec_fallback;
\&        long max_refresh;
\&        int proto;
\&    } val_global_opt_t;
.Ve
.PP
Setting a value of 1 for \fIlocal_is_trusted\fR is equivalent to specifying the
\&\fItrust-oob-answers\fR option in the \fBdnsval.conf\fR file to \fByes\fR. 
Similarly, setting the \fIclosest_ta_only\fR and \fIrec_fallback\fR
members to 1 is equivalent to setting the \fIclosest-ta-only\fR
and \fIrec-fallback\fR options to \fByes\fR in the \fBdnsval.conf\fR file. 
A value of 0 is equivalent to the \fBno\fR setting for these option.
.PP
Setting the \fIedns0_size\fR member to a particular value has the same effect 
setting the \fIedns0\-size\fR option in the \fBdnsval.conf\fR file.
.PP
Setting the \fImax_refresh\fR member to a particular value has the same effect 
setting the \fImax-refresh\fR option in the \fBdnsval.conf\fR file.
.PP
Setting the \fIproto\fR member to a particular value has the same effect 
setting the \fIproto\fR option in the \fBdnsval.conf\fR file.
.PP
\&\fIenv_policy\fR and \fIapp_policy\fR can be set to one of \fB\s-1VAL_POL_GOPT_DISABLE\s0\fR,
\&\fB\s-1VAL_POL_GOPT_ENABLE\s0\fR, or \fB\s-1VAL_POL_GOPT_OVERRIDE\s0\fR.  These values correspond
directly to the \fIdisable\fR, \fIenable\fR and \fIoverride\fR options for the
\&\fIenv-policy\fR and \fIapp-policy\fR settings in the \fBdnsval.conf\fR file.
.PP
Values of \-1 for the above fields are treated as \*(L"ignore\*(R" conditions
in that the validator will not include dynamically supplied global
options with a value of \-1 when creating its context.  This allows an
application to overriding a subset of global options while using the
global options supplied in the \fBdnsval.conf\fR file by default.
.PP
The log_target field enables the application to supply log targets 
\&\fIin addition\fR to the ones configured in the configuration file.
.PP
See \fB\f(BIdnsval.conf\fB\|(3)\fR for more details on specifying validator policy.
.PP
Default query flags can be set and unset for a given context using 
\&\fI\fIval_context_setqflags()\fI\fR.  This allows an application to alter the 
\&\s-1DNSSEC\s0 validator processing, while still having most of the more 
granular default configuration specified in its configuration file.  
The \fIaction\fR parameter can be set to one of the following.
.IP "\fB\s-1VAL_CTX_FLAG_SET\s0\fR" 4
.IX Item "VAL_CTX_FLAG_SET"
This option causes the validator to set the given flag as one of the default 
query flags.
.IP "\fB\s-1VAL_CTX_FLAG_RESET\s0\fR" 4
.IX Item "VAL_CTX_FLAG_RESET"
This option causes the validator to reset the given flag if it was set as one 
of the default query flags for the context.
.PP
Answers returned by \fI\fIval_resolve_and_check()\fI\fR are made available in the
\&\fI*results\fR linked list.  Each answer corresponds to a distinct RRset;
multiple RRs within the RRset are part of the same answer.  Multiple answers
are possible when \fItype\fR is \fIns_t_any\fR or \fIns_t_rrsig\fR.
.PP
Individual elements in \fI*results\fR point to \fIval_authentication_chain\fR linked
lists.  The authentication chain elements in \fIval_authentication_chain\fR
contain the actual RRsets returned by the name server in response to the query.
.PP
Validation result values returned in \fIval_result_chain\fR can be can be
converted into \s-1ASCII\s0 format using the \fI\fIp_val_status()\fI\fR.  Authentication chain
status values returned in each element of the \fIval_authentication_chain\fR
linked list can be can be converted into \s-1ASCII\s0 format using the
\&\fI\fIp_ac_status()\fI\fR functions.
.PP
While some applications such as \s-1DNSSEC\s0 troubleshooting utilities and packet
inspection tools may look at individual authentication chain elements to
identify the actual reasons for validation error, most applications will only
be interested in a single error code for determining the authenticity of data.
.PP
\&\fI\fIval_isvalidated()\fI\fR identifies if a given validation result status value
corresponds to an answer that was cryptographically verified and validated
using a locally configured trust anchor.
.PP
\&\fI\fIval_istrusted()\fI\fR identifies if a given validator status value is trusted.
An answer may be locally trusted without being validated.
.PP
\&\fI\fIval_does_not_exist()\fI\fR identifies if a given validator status value
corresponds to one of the non-existence types.
.PP
The \fIlibval\fR library internally allocates memory for \fI*results\fR and this
must be freed by the invoking application using the \fI\fIfree_result_chain()\fI\fR
interface.
.SH "DATA STRUCTURES"
.IX Header "DATA STRUCTURES"
.IP "\fIstruct val_result_chain\fR" 4
.IX Item "struct val_result_chain"
.Vb 10
\&  struct val_result_chain
\&  {
\&      val_status_t                     val_rc_status;
\&      char                            *val_rc_alias;
\&      struct val_rrset_rec            *val_rc_rrset;
\&      struct val_authentication_chain *val_rc_answer;
\&      int                              val_rc_proof_count;
\&      struct val_authentication_chain *val_rc_proofs[MAX_PROOFS];
\&      struct val_result_chain         *val_rc_next;
\&  };
.Ve
.RS 4
.IP "\fIval_rc_answer\fR" 4
.IX Item "val_rc_answer"
Authentication chain for a given RRset.
.IP "\fIval_rc_next\fR" 4
.IX Item "val_rc_next"
Pointer to the next RRset in the set of answers returned for a query.
.IP "\fIval_rc_proofs\fR" 4
.IX Item "val_rc_proofs"
Pointer to authentication chains for any proof of non-existence that were
returned for the query.
.IP "\fIval_rc_proof_count\fR" 4
.IX Item "val_rc_proof_count"
Number of proof elements stored in \fIval_rc_proofs\fR.  The number cannot exceed
\&\fB\s-1MAX_PROOFS\s0\fR.
.IP "\fIval_rc_alias\fR" 4
.IX Item "val_rc_alias"
For an val_result_chain element that points to a name alias, this field
contains the target value.
.IP "\fIval_rc_rrset\fR" 4
.IX Item "val_rc_rrset"
For a \fIval_result_chain\fR element that contains a valid (not \s-1NULL\s0)
\&\fIval_rc_answer\fR field, the \fIval_rc_rrset\fR field points to the top-most
\&\fIval_rrset_rec\fR element in the \fIval_rc_answer\fR authentication chain.
.IP "\fIval_rc_status\fR" 4
.IX Item "val_rc_status"
Validation status for a given RRset.  This can be one of the following:
.RS 4
.IP "\fB\s-1VAL_SUCCESS\s0\fR" 4
.IX Item "VAL_SUCCESS"
Answer received and validated successfully.
.IP "\fB\s-1VAL_NONEXISTENT_NAME\s0\fR" 4
.IX Item "VAL_NONEXISTENT_NAME"
No name was present and a valid proof of non-existence confirming the missing
name (\s-1NSEC\s0 or \s-1NSEC3\s0 span) was returned.  The components of the proof were also
individually validated.
.IP "\fB\s-1VAL_NONEXISTENT_TYPE\s0\fR" 4
.IX Item "VAL_NONEXISTENT_TYPE"
No type exists for the name and a valid proof of non-existence confirming the
missing name was returned.  The components of the proof were also individually
validated.
.IP "\fB\s-1VAL_NONEXISTENT_NAME_NOCHAIN\s0\fR" 4
.IX Item "VAL_NONEXISTENT_NAME_NOCHAIN"
No name was present and a valid proof of non-existence confirming the missing
name was returned.  The components of the proof were also identified to be
trustworthy, but they were not individually validated.
.IP "\fB\s-1VAL_NONEXISTENT_TYPE_NOCHAIN\s0\fR" 4
.IX Item "VAL_NONEXISTENT_TYPE_NOCHAIN"
No type exists for the name and a valid proof of non-existence confirming the
missing name (\s-1NSEC\s0 or \s-1NSEC3\s0 span) was returned.  The components of the proof
were also identified to be trustworthy, but they were not individually
validated.
.IP "\fB\s-1VAL_PINSECURE\s0\fR" 4
.IX Item "VAL_PINSECURE"
The record or some ancestor of the record in the authentication chain towards
the trust anchor was known to be provably insecure.
.IP "\fB\s-1VAL_PINSECURE_UNTRUSTED\s0\fR" 4
.IX Item "VAL_PINSECURE_UNTRUSTED"
The record or some ancestor of the record in the authentication chain towards
the trust anchor was known to be provably insecure. But the provably insecure
condition was configured as untrustworthy.
.IP "\fB\s-1VAL_BARE_RRSIG\s0\fR" 4
.IX Item "VAL_BARE_RRSIG"
No \s-1DNSSEC\s0 validation possible, query was for an \s-1RRSIG.\s0
.IP "\fB\s-1VAL_IGNORE_VALIDATION\s0\fR" 4
.IX Item "VAL_IGNORE_VALIDATION"
Local policy was configured to ignore validation for the zone from which this
data was received.
.IP "\fB\s-1VAL_UNTRUSTED_ZONE\s0\fR" 4
.IX Item "VAL_UNTRUSTED_ZONE"
Local policy was configured to reject any data received from the given zone.
.IP "\fB\s-1VAL_OOB_ANSWER\s0\fR" 4
.IX Item "VAL_OOB_ANSWER"
Answer was obtained using some Out Of Band method, such as a local
configuration file.
.IP "\fB\s-1VAL_BOGUS\s0\fR" 4
.IX Item "VAL_BOGUS"
Response could not be validated due to signature verification failures or the
inability to verify proofs in the authentication chain.
.IP "\fB\s-1VAL_DNS_ERROR\s0\fR" 4
.IX Item "VAL_DNS_ERROR"
Some error was encountered during \s-1DNS\s0 processing.
.IP "\fB\s-1VAL_INDETERMINATE\s0\fR" 4
.IX Item "VAL_INDETERMINATE"
Response could not be validated due to missing \s-1DNSSEC\s0 meta-data.
.IP "\fB\s-1VAL_NOTRUST\s0\fR" 4
.IX Item "VAL_NOTRUST"
All available components in the authentication chain verified properly, but
there was no trust anchor available.
.RE
.RS 4
.Sp
Status values in \fIval_status_t\fR returned by the validator can be displayed
in \s-1ASCII\s0 format using \fI\fIp_val_status()\fI\fR.
.RE
.RE
.RS 4
.RE
.IP "\fIstruct val_authentication_chain\fR" 4
.IX Item "struct val_authentication_chain"
.Vb 6
\&  struct val_authentication_chain
\&  {
\&      val_astatus_t                    val_ac_status;

dnssec-tools/NEWS
This file contains a list of major changes per release.  See the
ChangeLog file for a complete set of changes and their details
details.

2.2.2
 - A release that only merges a version fork between two upstreams

2.2.1
 - Bug fix release introduced by Wes Hardaker for Fedora packaging

2.2
 - Various bug fixes in multiple scripts and in the validation library
 - Note: The default zone file parser has switched to the
   Net::DNS::ZoneFile based parser, as the Net::DNS software is
   changing in non compatible ways and the Net::DNS::ZoneFile parser
   has caught up in speed to the Net::DNS::ZoneFile::Fast parser
   anyway.

2.1
 - New Features:
    - donuts:       - Added the ability to summarize information
                      about a zone in the output, such as the upcoming
                      entire zone expiry time, etc
                    - Added the ability to query live zones for
                      records to analyze.  EG:
                      donuts live:good-a,badsign-a test.dnssec-tools.org
                    - Added a -V switch to dump records analyzed
    - libval:       - Add support for conditionally checking all RRSIGs
                      on an assertion even if one that validates is
                      already found.
                    - Look for zonecuts based on NS records, not SOA
                    - Added initial support for TSIG in order to enable
                      libval to query recursive name servers that
                      authorized recursive lookup for only those hosts
                      that used a particular TSIG key.
    - Validator.pm  - Store respondent name server information in result
                      structure.
    - Owl           - additional sensor modules
                    - additional data analysis on manager
                    - logging to the Owl sensors modules
                    - optimized sensor data organization
                     (requires software upgrades on both sensor and
                      manager at the same time)
                    - added -restart option to owl-sensord for
                      restarting sensor modules
                    - improvements to the installation guide
    - rollerd       - generalized zonegroup entry in rollecs to be lists of tags
                    - rndc option support added
    - dnssec-check  - Ported to Qt5
    - dnssec-nodes  - Ported to Qt5
    - lookup        - Ported to Qt5
    - dnssec-system-tray
                    - Ported to Qt5

 - Bug Fixes
                    - Fixed bugs in libval, rollerd, blinkenlights, Owl
                      sensor modules, and Owl manager
                    - Use rlimits to try and limit file descriptor use in
                      libsres so we don't run out of available sockets.
                    - Eliminate a few hardcoded paths in various perl modules
                    - Fix various compiler warnings
                    - Update autoconf and related files


2.0
 - New Features
    - OWL           - The Owl Monitoring System uses timed DNS queries
                      to monitor basic network functionality.  The system
                      consists of a manager host and a set of sensor hosts.
                      The Owl sensors perform periodic DNS queries and
                      report to the Owl manager the time taken for each
                      query.  Over time, this shows the responsiveness of
                      the DNS infrastructure.
    - dnssec-nodes  - Many new features have been added:
                    - The validation tree now supports clicking on
                      boxes to highlight it and the arrows that derive
                      from it.  Great for use when teaching about
                      DNSSEC.
                    - An extensive filter/effect editor now lets you
                      tailor the look of a graph to color-code, set
                      the alpha levels, etc of nodes based on their
                      names, status, data types, etc.
                    - Right clicking on a node lets you center the
                      graph on that node.
                    - More data types are collected and shown in the
                      data view.
                    - Support for arguments on the command line for
                      parsing log files, pcap files and domain names.
                    - The validation view has received a visual clean-up
                    - Many other bug fixes
    - Bloodhound:   - A mozilla-based DNSSEC-enabled browser with DANE support
                    - Added support for validation of SSL certificates
                      using the DANE protocol.
    - curl          - Added support for validation of SSL certificates
                      using the DANE protocol.
    - libval        - Added support for local DANE validation
                    - Extended the dt-danechk commandline tool to check
                      the X509 cert provided over the SSL connection
                      against the TLSA record. 
                    - Optimized glue record lookup when the only ip
                      addresses configured for the host are for a single
                      address family (ipv4 or ipv6) 
                    - fine tune res_io source management
    - dnssec-check  - dnssec-check now checks DNAME support
    - rollerd       - A new set of steps for KSK rollover has been
                      implemented.  A cache-expiration wait phase has
                      been moved after the publication of DS records in
                      order to allow name caches to reflect the changes.
                      In addition to rollerd, supporting program have
                      been modified to recognize this change.
    - rollrec files - A new "information rollrec" has been added to the
                      rollrec files.  This will allow infomration to be

dnssec-tools/tools/convertar/lib/Net/DNS/SEC/Tools/TrustAnchor/Libval.pm
package Net::DNS::SEC::Tools::TrustAnchor::Libval;

use strict;
use Net::DNS::SEC::Tools::TrustAnchor;

our @ISA = qw(Net::DNS::SEC::Tools::TrustAnchor);
our $VERSION = '0.1';

use XML::Simple;

sub init_extras {
    my $self = shift;
    # XXX: allow for other contexts besides :
    $self->{'header'} = ": trust-anchor\n";
    $self->{'trailer'} = ";\n";
}

sub read_content {
    my ($self, $location, $options) = @_;

    $location ||= $self->{'file'};
    $options ||= $self->{'options'};

    my $doc = { delegation => {}};

    my $intrustanchor = 0;
    open(I, "<", $location);
    while (<I>) {
	if (/^\s*# EIVER/) {
	    my $localinfo =
	      $self->parse_extra_info_string($_, $doc, "#");
	}

	next if (/^\s*#/);

	if (/^(\S+)\s+trust-anchor/) {
	    $intrustanchor = 1;
	} elsif ($intrustanchor && /^\s*(\S+)\s+(DNSKEY|)\s*"([^"]+)"/) {
	    my $name = $1;
	    my $data = $3;
	    my ($flags, $alg, $digesttype, $content) =
	      ($data =~ /^(\d+)\s+(\d+)\s+(\d+)\s+(.*)/);
	    $content =~ s/ //g;
	    push @{$doc->{'delegation'}{$name}{'dnskey'}},
	      {
	       flags => $flags,
	       algorithm => $alg,
	       digesttype => $digesttype,
	       content => $content,
	      };
	    $intrustanchor = 0 if (/;/);
	} elsif ($intrustanchor && /^\s*(\S+)\s+DS\s+"([^"]+)"/) {
	    my $name = $1;
	    my $data = $2;
	    my ($keytag, $alg, $digesttype, $content) =
	      ($data =~ /^(\d+)\s+(\d+)\s+(\d+)\s+(.*)/);
	    $content =~ s/ //g;
	    push @{$doc->{'delegation'}{$name}{'ds'}},
	      {
	       keytag => $keytag,
	       algorithm => $alg,
	       digesttype => $digesttype,
	       content => $content,
	      };
	    $intrustanchor = 0 if (/;/);
	} elsif (/;/) {
	    $intrustanchor = 0;
	}
    }

    return $doc;
}

sub write_ds {
    my ($self, $fh, $name, $record) = @_;
    my $status;
    $fh->printf("\t%15s DS \"$record->{keytag} $record->{algorithm} $record->{digesttype} $record->{content}\"\n", $name);
}

sub write_dnskey {
    my ($self, $fh, $name, $record) = @_;
    my $status;
    $fh->printf("\t%15s \"$record->{flags} $record->{algorithm} $record->{digesttype} $record->{content}\"\n", $name);
}

sub write_trailer {
    my ($self, $fh, $options, $data) = @_;
    if ($self->{'options'}{'write_expectations'}) {
	$fh->printf(";\n\n");
	$fh->printf(": zone-security-expectation\n");
	$fh->printf("    %-50.50s ignore\n", ".");
	foreach my $zone (keys(%{$data->{'delegation'}})) {
	    $fh->printf("    %-50.50s validate\n", $zone);
	}
	$fh->printf(";\n");
    }
}

=pod

=cut

